src/cargo/util/graph.rs

```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
     }
 }
 
-impl<N: Eq + Ord + Clone, E: Eq + Clone> PartialEq for Graph<N, E> {
+impl<N: Eq + Hash + Clone, E: Eq + Clone + PartialEq> PartialEq for Graph<N, E> {
     fn eq(&self, other: &Graph<N, E>) -> bool {
         self.nodes.eq(&other.nodes)
     }
 }
-impl<N: Eq + Ord + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
+impl<N: Eq + Hash + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
 
-impl<N: Eq + Ord + Clone, E: Clone> Clone for Graph<N, E> {
-    fn clone(&self) -> Graph<N, E> {
-        Graph {
-            nodes: self.nodes.clone(),
+/// This is a directed Graph structure, that builds on the `Graph`'s "append only" internals
+/// to provide:
+///  - `O(ln(number of clones))` clone, no dependence on the size of the graph
+///  - the clone has no overhead to read the `Graph` as it was
+///  - no overhead over using the `Graph` directly when modifying the biggest clone
+/// Is this too good to be true?
+///  - If a modification (`&mut` method) is done to a smaller older clone then a full `O(N)`
```

Comment:
stm this is a bit of a performance footgun. I think it would be better for such a modification to fail and have an explicit deep clone method.